flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java

```diff
@@ -164,6 +168,27 @@ void addCreditOrResumeConsumption(
         }
     }
 
+    /**
+     * Announces remaining backlog to the consumer after the available data notification or data
+     * consumption resumption.
+     */
+    private void announceBacklog(NetworkSequenceViewReader reader) {
+        int backlog = reader.getRemainingBacklog();
+        if (backlog > 0) {
```

Comment:
The backlog can be 0. For example, after resumption from checkpoint, if there is no pending data in the subpartition.

Reply:
Have you tested this @wsry ? After all it seems like if reader is available, it should have `backlog > 0`, shouldn't it?

Reply:
Secondly, `getRemainingBacklog()` is a very costly operation (additional synchronisation) that I think could have been avoided:

1. Backlog can go up only as a result of `org.apache.flink.runtime.io.network.netty.PartitionRequestQueue#notifyReaderNonEmpty(reader)`. 
2. Backlog can go down, only as a result of polling the data from the `reader`.
3. So instead of using thread safe `reader.getRemainingBacklog()`, we could re-use existing synchronisation in 1. and 2., to maintain `remainingBacklog` in the netty thread (in the `PartitionRequestQueue`.


But it would be even better to completely avoid this check (my comment above).

Reply:
You are right, we can do some optimization here.

Reply:
I think there is one thing blocking us from reusing the backlog from PartitionRequestQueue#notifyReaderNonEmpty(reader): we have duplicated availability notifications. We may get outdated backlog.

Reply:
Maybe we can reuse the return value of is available. Instead of just return a bool flag, we may also return the backlog. I will give it a try.

Reply:
`NetworkSequenceViewReader#isAvailable()` has the same problem of synchronisation costs.

> I think there is one thing blocking us from reusing the backlog from PartitionRequestQueue#notifyReaderNonEmpty(reader): we have duplicated availability notifications. We may get outdated backlog.

What is the problem?  Keep in mind that spurious notifications shouldn't be that big of a problem. If we sometimes wake up too many times, and we rarely send incorrect backlog information that should be fine, as long as we always over estimate the backlog.

After all even in your version I think there can be a race condition where `PartitionRequestQueue` is notified reader is non empty, you check the `getRemainingBacklog()` and send the `BacklogAnnouncement` message, while before it gets processed by the receiver, some buffer is polled from this reader and backlog goes down to `0`. And as a result, receiver assigns as a credit to a sender that doesn't have any data anymore?

Reply:
NetworkSequenceViewReader#isAvailable() has already been called by PartitionRequestQueue#enqueueAvailableReader:
```
private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
if (reader.isRegisteredAsAvailable() || !reader.isAvailable()) {
return;
}
......
}
```
I think we can change it to:
```
    private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
        if (reader.isRegisteredAsAvailable()) {
            return;
        }

        ResultSubpartitionView.AvailabilityWithBacklog availabilityWithBacklog =
                reader.getAvailabilityAndBacklog();
        if (!availabilityWithBacklog.isAvailable()) {
            int backlog = availabilityWithBacklog.getBacklog();
            if (backlog > 0 && reader.needAnnounceBacklog()) {
                announceBacklog(reader, backlog);
            }
            return;
        }
......
}
```
What do you think?

> After all even in your version I think there can be a race condition where PartitionRequestQueue is notified reader is non empty, you check the getRemainingBacklog() and send the BacklogAnnouncement message, while before it gets processed by the receiver, some buffer is polled from this reader and backlog goes down to 0. And as a result, receiver assigns as a credit to a sender that doesn't have any data anymore?

I think there is no such problem. One reason is that ```getRemainingBacklog()``` is in netty thread and we can guarantee the order, the downstream will always process the ```BacklogAnnouncement``` before the buffer. The other reason is that we only announce credit when there is no credit available which means if there is credit available, only buffer with backlog will be sent, if there is no credit available, only ```BacklogAnnouncement``` will be send, buffer will be always wait for credit. It is also not a problem if we announce the same credit twice because we can guarantee the process order.

Reply:
Yes you are right. `reader.isAvailable()` is already being called. I'm not sure why I've missed that.

`reader.getAvailabilityAndBacklog();` should work fine :)